atc/atccmd/command.go

```diff
@@ -1187,8 +1188,8 @@ func (cmd *RunCommand) constructLockConn(driverName string) (*sql.DB, error) {
 		return nil, err
 	}
 
-	dbConn.SetMaxOpenConns(1)
```

Comment:
The lock connection has to be forced to 1 because we use session-level locks, which are associated to a connection. If we pool them, locks will potentially be lost when a connection in the pool is closed.

I don't think this would be the source of a lot of churn; it should just be one connection that stays open the whole time. :thinking: We never checked though.

Reply:
This is actually the source of the first bottleneck.  If you look at the first set of calls in #4200 you'll see that AcquireLock is the dominate call.  This turns out to be because there is a race condition between giving up the lock from a previous call, getting a connection to the free list, and establishing new connections.  We found that because the connection pool manager is operating in a separate thread, the system was never able to pull from the idle connection list, so it would establish a new connection.  Then it had hit the max open and had to free the existing.

Reply:
I'm surprised it would establish a new connection and do all the TLS stuff even when it knows it's reached `MaxOpenConns` (1), though. I would have expected it to just wait for the open one to go back to idle or something.

Looking at #4200, I see `AcquireSchedulingLock` but not `AcquireLock`. The time could just as easly be coming from this `Exec` call, which also shows up in the profile but is using the normal connection pool, not the lock pool:

https://github.com/concourse/concourse/blob/ca0fd4cb1ed0b619fd1e083b5b3c827ffb669e63/atc/db/pipeline.go#L917-L922

I see that you confirmed the fix was effective, but I wonder if it would have been just as effective without changing the lock connection settings. Maybe it's really just a matter of bumping the idle connections for the other connection pools?

If not, and the lock pool is really the culprit, we'll need to find a way to fix it without bumping the pool, because it really needs to be a single dedicated connection for the locking semantics to work correctly.

The code that uses the lock connection is here: https://github.com/concourse/concourse/blob/master/atc/db/lock/lock.go

Maybe it should use a [`driver.Conn`](https://golang.org/pkg/database/sql/driver/#Conn) instead? That would avoid the pooling altogether. We'd need a way to re-establish the connection if it breaks, though.

Reply:
Here is the trace from the original code: [https://drive.google.com/file/d/1Pobv2FcULJykqHq5nsaMYRZsX3hGCSZ6/view?usp=sharing](5.4.0), and here is the trace with the changes: [https://drive.google.com/file/d/1Ep3KM9X4Cpc4cCXmpFAB0_c5KK1XrgQG/view?usp=sharing](5.4.0+#4232)

You can see the delta.  I first concentrated on the issue of the `AcquireSchedulingLock `(misnamed above by me).  Once I removed that as a bottleneck, the actual queries came up as an issue.  How this is manifesting on the system is that we end up with well over 100 (sometimes as high as 200) TCP connections in TIME_WAIT, which means they have been closed and the system is waiting to allow the port to be available again.  

I think your idea of using the `driver.Conn`is probably the right one, but I have to wonder whether it makes more sense to actually use a single pool and set the max Idle to be a reasonably high number (20-40).  The re-use ratio should be much higher and the connection pooling should start working as designed.  I understand that the default configuration of Postgres is to limit the number of connections to 100, but with multiple webs, and the backoff methodology already in place, I think that it doesn't really matter.  The goal should be to have the maximum number of requests serviced within the limits of the environment, and in a high workload environment, you will have to raise that limit on Postgres.

Reply:
Looking at the graph, it looks like it's all coming from the non-lock connection pool. The graph shows the cost going through a `*countingConn` which is only added here:

https://github.com/concourse/concourse/blob/cfc598920835c03f8c3b7c68dd60f03b6e14cc2c/atc/atccmd/command.go#L1166

We can't use a pool of any sort because the nature of the queries we use to acquire/release. When releasing, the query has to be run on the same connection that acquired the lock. We use advisory locks and avoid holding transactions open, which we've observed causes an increase in database CPU usage the longer they're held open. Our solution was to just have a single dedicated connection for locks.

Reply:
I thought (admittedly, I didn't look as closely as you have) that there were two sets of mutexes around the calls to `AcquireSchedulingLock`.  Why does the actual connection matter at that point?  The connection does not actually create the lock in any real way does it?  The Mutex is what actually protects the ability to update/hold the lock.  What I saw was that the connection was not actually being re-used anyway.  As I understood it, the `*countingConn.Exec`  is the wrapper around all of the DB connection requests.  The size of the timings going into it, show that initially the big issue was around   `AcquireSchedulingLock`.  Once I cleared that it moved as the core issue was really around the number of maxIdleConnections.

Reply:
The lock is created using the connection here:

https://github.com/concourse/concourse/blob/12b4d7cfc178e0f0639e4b5b899b4d2f3784b012/atc/db/lock/lock.go#L213

The additional locking is just to prevent the same `web` node from acquiring the same lock twice, since Postgres will allow the same connection to acquire the same lock multiple times as long as it's released each time. This isn't what we want; we want it to wait on the original lock to be released.

`*countingConn` is only used for the non-lock connections; there's no need to count connections for the lock connection as we enforce that there is always only one anyway.

The fact that `AcquireSchedulingLock` showed up in the profile so much might just be because that operation happens once every 10 seconds for every pipeline, and it's the first database operation performed on that interval. It might also be spiky, i.e. many pipelines are being scheduled around the same time, which would lead to a bunch of connections being created. Increasing the max idle conns beyond the default of 2 as you did would help ease that spikiness, but I think it's just a coincidence that it was a '*Lock' call, and that the issue doesn't lie in the lock connection pool.

Reply:
As an aside, the actual observations are that we are bouncing between 21 and 48 open connections to the DB.  I'm still struggling to understand why treating the connection pools as independent is buying anything here.  

The reason I believe that the lock connection pool is important is that it is used frequently, and the mutexes around it guarantee that there is a backlog of requests coming in to capture the next lock.  This seems to result in the problem where there are no idle connections yet on the queue, and with the maxOpenConnections and maxIdleConnections set to 1, it was causing the situation where the pool manager had no choice but to establish a new connection for every request.  That's not good.  A pool size of 2 may be enough, but being set at 1 guarantees a race condition that results in re-establishing connections.

Reply:
As far as I can tell, if you have max connections set to 1, it's never going to open another connection. That would directly go against the documentation.

This is pretty easy to verify. Here's a gist demonstrating the behavior: https://gist.github.com/vito/55966975903a77ec31c0f36ace15fcc0

If you have `MaxOpen` set to 1 and something needs a new connection, it waits for the connection to become available again. It's not going to make a new connection; the limit is set to 1, and that would be 2 open connections.

Reply:
Thanks. I dug through the code, played with the example and I agree with the conclusion.  I've moved the lock back to it's original state, and just made the the two changes: 1.) allow MaxOpenConnections to be configurable via a flag, with same current default and 2.) set MaxIdleConnections to be 1/2 of MaxOpenConnections.  This should fix the issue of reconnections being a major issue, and allow for changes for particular environments.  I'm testing again on my side right now.